src/main/java/org/springframework/integration/kafka/outbound/KafkaProducerMessageHandler.java

```diff
@@ -193,9 +283,51 @@ protected void handleMessageInternal(final Message<?> message) throws Exception
 			headers = new RecordHeaders();
 			this.headerMapper.fromHeaders(message.getHeaders(), headers);
 		}
-		ProducerRecord<K, V> producerRecord = new ProducerRecord<K, V>(topic, partitionId, timestamp, (K) messageKey,
-				payload, headers);
-		ListenableFuture<?> future = this.kafkaTemplate.send(producerRecord);
+		final ProducerRecord<K, V> producerRecord = new ProducerRecord<K, V>(topic, partitionId, timestamp,
+				(K) messageKey, payload, headers);
+		ListenableFuture<?> future = null;
+		final SettableListenableFuture<RecordMetadata> finalFuture;
+		if (this.needsCallback) {
+			if (this.sync) {
+				finalFuture = new SettableListenableFuture<RecordMetadata>();
+				future = finalFuture;
+			}
+			else {
+				finalFuture = null;
+			}
+			this.kafkaTemplate.send(producerRecord, new Callback() {
+
+				@Override
+				public void onCompletion(RecordMetadata metadata, Exception exception) {
+					if (exception != null) {
+						if (finalFuture != null) {
+							finalFuture.setException(exception);
+						}
+						if (getSendFailureChannel() != null) {
+							Attributes attributes = new Attributes();
+							attributes.setAttribute(KafkaMessageDrivenChannelAdapter.KAFKA_RAW_DATA,
+									producerRecord);
+							Message<?> errorMessage = KafkaProducerMessageHandler.this.errorMessageStrategy
+									.buildErrorMessage(new MessageHandlingException(message, exception), attributes);
+							KafkaProducerMessageHandler.this.messagingTemplate.send(getSendFailureChannel(),
+									errorMessage);
+						}
+					}
+					else {
+						if (finalFuture != null) {
+							finalFuture.set(metadata);
+						}
+						if (getSendSuccessChannel() != null) {
+							KafkaProducerMessageHandler.this.messagingTemplate.send(getSendSuccessChannel(), message);
```

Comment:
Why do we send a `requestMessage` without any `RecordMetadata` options?
Maybe we really should have it as a `AbstractMessageProducingHandler` and send the real reply instead of an extra `sendSuccessChannel`?